Lib/urllib/request.py

```diff
@@ -937,7 +937,7 @@ class AbstractBasicAuthHandler:
 
     # allow for double- and single-quoted realm values
     # (single quotes are a violation of the RFC, but appear in the wild)
-    rx = re.compile('(?:.*,)*[ \t]*([^ \t]+)[ \t]+'
+    rx = re.compile('(?:[^,]*,)*[ \t]*([^ \t]+)[ \t]+'
```

Comment:
`(?:.*,)*` is equivalent to `(?:.*,)?`.

But since this regular expresion is only used with `search()`. `(?:.*,)*[ \t]*` can be removed at all.

I'll analyze whether it is correct or there is an error in the regular expression.

Reply:
Well, I am cannot say that I completely understand the code, but to give it some sense we can either

1. Replace `rx.search()` with `rx.match()` and replace `(?:.*,)*` with `(?:.*,)?`.

or

2. Keep `rx.search()` and replace `(?:.*,)*` with `(?:^|,)`.

Do not keep `(?:[^,]*,)*`. It is a waster of resources.

Reply:
Humm, options 1 and 2 are not equivalent if the field value contains more than one challenge. Option 2 is closer to the current behavior. But correct support of more than one challenge need rewriting the code.

https://tools.ietf.org/html/rfc7235#section-4.1

Reply:
The current patch seems to give an O(n^3) time to evaluate - much better than O(2^n), but still very slow - with 2000 commas it takes about a minute to evalute. With 65000 it takes much, much longer. Testing using the code from [here](https://bugs.python.org/msg356785) gave the following (commas, seconds) values:
`[(100, 0.124), (250, 0.261), (500, 0.923), (750, 2.85), (1000, 6.433), (1250, 12.608), (1500, 21.576), (2000, 50.751)]`

Reply:
I was wrong, the first option is equivalent to the current behavior (returns the last realm).